snippets/c_api/iterate_events.cc

```diff
@@ -0,0 +1,181 @@
+void SetPairs(AlohalyticsKeyPairsEvent const * event, std::vector<char const *> & pairs)
+{
+  int i = 0;
+  for (std::pair<std::string, std::string> const & kp : event->pairs)
+  {
+    pairs[i++] = kp.first.c_str();
+    pairs[i++] = kp.second.c_str();
+  }
+}
+
+// Helper function for setting up geo coords
+void SetGeo(alohalytics::Location const kLocation, UserInfo & userInfo)
+{
+  if (kLocation.HasLatLon()) {
+    userInfo.lat = (float)kLocation.latitude_deg_;
+    userInfo.lon = (float)kLocation.longitude_deg_;
+  }
+}
+
+/*
+Helper function for setting up os type as a num for use in Python
+NOTE: C char will convert into python string - so we do not want that
+*/
+void SetOSType(char osType, UserInfo& userInfo)
+{
+  switch (osType)
+  {
+    case 'A':
+    {
+      userInfo.os_t = 1;
+      break;
+    }
+    case 'I':
+    {
+      userInfo.os_t = 2;
+      break;
+    }
+  };
+}
+
+std::string CompressUID(std::string uid)
+{
+  uid.erase(0, 2);
+  // TODO: think about just using fixed positions instead of more general attempt with searching
+  uid.erase(std::remove(uid.begin(), uid.end(), '-'), uid.end());
+  return uid;
+}
+
+/*
+This is public function used in Python as a main loop
+NOTE: We need a C interface, so prepare yourself for C-style structures in the params of
+this function (called from a Python code with a ctypes wrapper) and
+callback function (actual Python function - also a ctypes wrapper)
+*/
+
+DLLEXPORT void iterate(callbackFunction callback, char const ** kEventNames, int eventNamesNum)
+{
+  std::set<std::string> processKeys;
+  for (int i = 0; i < eventNamesNum; i++)
+  {
+    processKeys.insert(kEventNames[i]);
+  }
+
+  AlohalyticsIdServerEvent const * kPrevious = nullptr;
+  alohalytics::Processor(
+      [&kPrevious, callback, &processKeys](AlohalyticsIdServerEvent const * kIdEvent, AlohalyticsKeyEvent const * kEvent)
+      {
+
+        if (!processKeys.empty() && !processKeys.count(kEvent->key))
+        {
+          return;
+        }
+
+        EventTime const kEventTime = {
+          .client_created = kEvent->timestamp,
+          .server_upload = kIdEvent->server_timestamp
+        };
+
+        UserInfo userInfo = {
+          .os_t = 0,
+          .lat = 0.0,
+          .lon = 0.0
+        };
+
+        SetOSType(kIdEvent->id[0], userInfo);
+
+        std::memcpy(userInfo.raw_uid, CompressUID(kIdEvent->id).c_str(), 32);
+
+        char const * kKey = kEvent->key.c_str();
+
+        AlohalyticsKeyValueEvent const * kve = dynamic_cast<AlohalyticsKeyValueEvent const *>(kEvent);
+        if (kve)
+        {
+          char const * kValues[1] = {kve->value.c_str()};
+          callback(kKey, kEventTime, userInfo, kValues, 1);
+          return;
+        }
+
+        AlohalyticsKeyPairsEvent const * kpe = dynamic_cast<AlohalyticsKeyPairsEvent const *>(kEvent);
+        if (kpe)
+        {
+          std::vector<char const *> pairs(kpe->pairs.size() * 2);
```

Comment:
It's better to move proper vector size calculations to SetPair method. See my comment above.